src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs

```diff
@@ -92,50 +118,65 @@ public void GetProxyForUrl(SafeWinHttpHandle sessionHandle, Uri uri, out Interop
+                    out proxyInfo))
                 {
-                    repeat = false;
+                    WinHttpTraceHelper.Trace("WinInetProxyHelper.GetProxyForUrl: Using autoconfig proxy settings");
+                    useProxy = true;
+                    
+                    break;
                 }
                 else
                 {
                     var lastError = Marshal.GetLastWin32Error();
-                    if (lastError == Interop.WinHttp.ERROR_WINHTTP_AUTODETECTION_FAILED)
+                    WinHttpTraceHelper.Trace("WinInetProxyHelper.GetProxyForUrl: error={0}", lastError);
+
+                    if (lastError == Interop.WinHttp.ERROR_WINHTTP_LOGIN_FAILURE)
                     {
-                        // Fall back to manual settings if available.
-                        if (!string.IsNullOrEmpty(Proxy))
+                        if (repeat)
                         {
-                            proxyInfo.AccessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_NAMED_PROXY;
-                            proxyInfo.Proxy = Marshal.StringToHGlobalUni(Proxy);
-                            proxyInfo.ProxyBypass = string.IsNullOrEmpty(ProxyBypass) ? IntPtr.Zero : Marshal.StringToHGlobalUni(ProxyBypass);
+                            // We don't retry more than once.
+                            break;
                         }
                         else
                         {
-                            proxyInfo.AccessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_NO_PROXY;
-                            proxyInfo.Proxy = IntPtr.Zero;
-                            proxyInfo.ProxyBypass = IntPtr.Zero;
+                            repeat = true;
+                            autoProxyOptions.AutoLoginIfChallenged = true;
                         }
-
-                        repeat = false;
-                    }
-                    else if (lastError == Interop.WinHttp.ERROR_WINHTTP_LOGIN_FAILURE)
-                    {
-                        if (repeat)
-                        {
-                            throw WinHttpException.CreateExceptionUsingError(lastError);
-                        }
-
-                        repeat = true;
-                        autoProxyOptions.AutoLoginIfChallenged = true;
                     }
                     else
                     {
-                        WinHttpException.ThrowExceptionUsingLastError();
+                        break;
                     }
                 }
             } while (repeat);
+
+            // Fall back to manual settings if available.
+            if (!useProxy && !string.IsNullOrEmpty(Proxy))
+            {
+                proxyInfo.AccessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_NAMED_PROXY;
+                proxyInfo.Proxy = Marshal.StringToHGlobalUni(Proxy);
+                proxyInfo.ProxyBypass = string.IsNullOrEmpty(ProxyBypass) ?
+                    IntPtr.Zero : Marshal.StringToHGlobalUni(ProxyBypass);
```

Comment:
This code was always there, but just moved around in this method.

The caller, `WinHttpHandler.SetRequestHandleProxyOptions()` frees those native strings.